### PR TITLE
Realtime rc backoff

### PR DESCRIPTION
**Pull request by qdpham13**
Change exponential backoff logic to use base RC's backoff logic and persist backoff state to disk.

### CODE REVIEW

File: FirebaseRemoteConfig/Sources/RCNConfigRealtime.m [hunk 1]
```diff
@@ -612,13 +609,16 @@ - (void)URLSession:(NSURLSession *)session
                     task:(NSURLSessionTask *)task
     didCompleteWithError:(NSError *)error {
   _isRequestInProgress = false;
+  [self->_settings updateRealtimeExponentialBackoffTime:self->_remainingRetryCount == gMaxRetries];
```
Comment (karenyz) [inline FirebaseRemoteConfig/Sources/RCNConfigRealtime.m:612]: not sure that we need to update backoff time here?

File: FirebaseRemoteConfig/Sources/RCNConfigRealtime.m [hunk 2]
```diff
@@ -601,7 +599,6 @@ - (void)URLSession:(NSURLSession *)session
   } else {
     /// on success reset retry parameters
     _remainingRetryCount = gMaxRetries;
-    _retrySeconds = arc4random_uniform(5) + 1;
```
[the remark this answers is not in the document]
Comment (qdpham13) [reply FirebaseRemoteConfig/Sources/RCNConfigRealtime.m:604]: So we don't need to reset the backoff time here b/c this delegate is called when a response has been received and the connection is open so the backoff time is already < current time. We do need to reset the retry count here though and I updated that.

File: FirebaseRemoteConfig/Sources/RCNConfigRealtime.m [hunk 3]
```diff
@@ -358,13 +354,12 @@ - (void)retryHTTPConnection {
   dispatch_async(_realtimeLockQueue, ^{
     __strong RCNConfigRealtime *strongSelf = weakSelf;
     if (strongSelf->_remainingRetryCount > 0) {
-      if (strongSelf->_remainingRetryCount < gMaxRetries) {
-        double RETRY_MULTIPLIER = arc4random_uniform(3) + 2;
-        strongSelf->_retrySeconds *= RETRY_MULTIPLIER;
-      }
+      NSTimeInterval now = [[NSDate date] timeIntervalSince1970];
+      NSTimeInterval backOffTime = self->_settings.realtimeExponentialBackoffThrottleEndTime - now;
```
[the remark this answers is not in the document]
Comment (qdpham13) [reply FirebaseRemoteConfig/Sources/RCNConfigRealtime.m]: Updated.

File: FirebaseRemoteConfig/Sources/RCNConfigRealtime.m [hunk 4]
```diff
@@ -580,8 +575,11 @@ - (void)URLSession:(NSURLSession *)session
   _isRequestInProgress = false;
   NSHTTPURLResponse *_httpURLResponse = (NSHTTPURLResponse *)response;
   NSInteger statusCode = [_httpURLResponse statusCode];
+  [self->_settings updateRealtimeExponentialBackoffTime:self->_remainingRetryCount == gMaxRetries];
+
   if (statusCode != kRCNFetchResponseHTTPStatusOk) {
     [self pauseRealtimeStream];
+
     if ([self isStatusCodeRetryable:statusCode]) {
       [self retryHTTPConnection];
```
[the remark this answers is not in the document]
Comment (qdpham13) [reply FirebaseRemoteConfig/Sources/RCNConfigRealtime.m:584]: The backoff time is updated a few lines before hand on line 579

File: FirebaseRemoteConfig/Sources/RCNConfigRealtime.m [hunk 5]
```diff
@@ -131,9 +130,6 @@ - (instancetype)init:(RCNConfigFetch *)configFetch
     _options = options;
     _namespace = namespace;
 
-    // Set retry seconds to a random number between 1 and 6 seconds.
-    _retrySeconds = arc4random_uniform(5) + 1;
-
     _remainingRetryCount = gMaxRetries;
```
[the remark this answers is not in the document]
Comment (qdpham13) [reply FirebaseRemoteConfig/Sources/RCNConfigRealtime.m]: Yeah we do, good point. I updated it to persist the count too.

File: FirebaseRemoteConfig/Sources/RCNConfigRealtime.m (hunk 1 above)
Comment (qdpham13) [reply FirebaseRemoteConfig/Sources/RCNConfigRealtime.m]: This error occurs when the connection is interrupted client side do I think we should update the backoff time here too.

**Comment by qdpham13** [comment]
> overall i'm finding it hard to follow where the backoff is being reset/extended, left some comments but might be worth discussing more offline

The rule of thumb I was following is that if the connection attempt fails then update the backoff time to some future date and persist both backoff time and retry count. The only thing that needs to be reset is the retry count when there is a successful connection; we don't need to update the time interval b/c if a connection attempt happens then the backoff time is already less then the current time and it won't prevent future backoffs if it stays the same. Lmk if you would want to grab some more time and talk about it

File: FirebaseRemoteConfig/Sources/RCNConfigRealtime.m [hunk 6]
```diff
@@ -358,13 +354,12 @@ - (void)retryHTTPConnection {
   dispatch_async(_realtimeLockQueue, ^{
     __strong RCNConfigRealtime *strongSelf = weakSelf;
     if (strongSelf->_remainingRetryCount > 0) {
-      if (strongSelf->_remainingRetryCount < gMaxRetries) {
-        double RETRY_MULTIPLIER = arc4random_uniform(3) + 2;
-        strongSelf->_retrySeconds *= RETRY_MULTIPLIER;
-      }
+      NSTimeInterval backOffTime = self->_settings.getRealtimeBackoffInterval;
```
Comment (karenyz) [inline FirebaseRemoteConfig/Sources/RCNConfigRealtime.m:357]: nit: `backoffInterval` to indicate it's an interval not a timestamp

**Review by karenyz — APPROVED** [review]
